bluepysnap/nodes.py

```diff
@@ -385,89 +370,17 @@ def _get_node_set(self, node_set_name):
-            node_ids = node_ids if node_ids else self._data.index.values
-            node_ids = np.intersect1d(node_ids, self.ids(node_set))
-        return queries, self._positional_mask(node_ids)
-
-    def _properties_mask(self, queries, raise_missing_prop):
-        """Return mask of node IDs with rows matching `props` dict."""
-        # pylint: disable=assignment-from-no-return
-        circuit_keys = {POPULATION_KEY, NODE_ID_KEY, NODE_SET_KEY}
-        unknown_props = set(queries) - set(self._data.columns) - circuit_keys
-        if unknown_props:
-            if raise_missing_prop:
-                raise BluepySnapError(
-                    "Unknown node properties: [{0}]".format(", ".join(unknown_props)))
-            return np.full(len(self._data), fill_value=False)
-
-        queries, mask = self._circuit_mask(queries)
-        if not mask.any():
-            # Avoid fail and/or processing time if wrong population or no nodes
-            return mask
-
-        for prop, values in queries.items():
-            prop = self._data[prop]
-            if np.issubdtype(prop.dtype.type, np.floating):
-                v1, v2 = values
-                prop_mask = np.logical_and(prop >= v1, prop <= v2)
-            elif isinstance(values, str) and values.startswith('regex:'):
-                prop_mask = _complex_query(prop, {'$regex': values[6:]})
-            elif isinstance(values, Mapping):
-                prop_mask = _complex_query(prop, values)
-            else:
-                prop_mask = np.in1d(prop, values)
-            mask = np.logical_and(mask, prop_mask)
-        return mask
-
-    def _operator_mask(self, queries, raise_missing_prop):
-        """Handle the query operators '$or', '$and'."""
-        if len(queries) == 0:
-            return np.full(len(self._data), True)
-
-        # will pop the population and or/and operators so need to copy
-        queries = deepcopy(queries)
-        first_key = list(queries)[0]
-        if first_key == '$or':
-            queries = queries.pop("$or")
-            operator = np.logical_or
-        elif first_key == '$and':
-            queries = queries.pop("$and")
-            operator = np.logical_and
-        else:
-            return self._properties_mask(queries, raise_missing_prop)
+    def _resolve_nodesets(self, queries):
+        def _resolve(queries, queries_key):
+            if queries_key == query.NODE_SET_KEY:
+                if query.AND_KEY not in queries:
+                    queries[query.AND_KEY] = []
+                queries[query.AND_KEY].append(self._get_node_set(queries[queries_key]))
+                del queries[queries_key]
 
-        mask = np.full(len(self._data), first_key != "$or")
-        for query in queries:
-            mask = operator(mask, self._operator_mask(query, raise_missing_prop))
-        return mask
+        resolved_queries = deepcopy(queries)
+        query.traverse_queries_bottom_up(resolved_queries, _resolve)
```

Comment:
The more I read your code the more I think having a Query class would make sense. 
What do you think ?

Reply:
It is possible, I need to look into it. I kept old code as much as possible, so new changes are easier to see.

Reply:
After some experiments. There is one problem. Such class must be used only within `.query` module. We will need to transform `queries` dict into an instance of this class. Such transformation would require `queries` traversal. In order to resolve ids, we would need to traverse the instance of this class. So, double traversal. I'm against double traversal. Possibly I picked the wrong class representation. Please suggest your alternatives:

```
class QueryTree:
    def __init__(self, query_dict, parent=None):
        self._parent = parent
        if parent is None:
            self._key = AND_KEY
            self._value = None
            self._children = []
            for key, value in query_dict.items():
                self._children.append(QueryTree({key: value}, self))
        else:
            self._key, self._value = unfold(query_dict)
            if self.key == OR_KEY:
                pass
            elif self.key == AND_KEY:
                pass
            elif regex_case:
                pass
```

Reply:
ah I did not think about that indeed ... 

I spent some times thinking about it and I don t have anything really clean in my mind. Also because the traverse function aims at transforming the query but without any expected content at the end (can be a dict with other queries or a dict with np arrays or whatever). 

Hum yes this is not that easy to have a good data structure... so let s keep it like this and we can make another path on this if we have a brilliant idea at some point.